File: Baekjoon/1/1325.cpp
#include <stdio.h>
#include <vector>
#include <stack>

#define MAX 100001

int n, m;
int ans[MAX];
std::vector<int> adj[MAX];

int dfs(int v)
{
    bool visited[MAX] = {false};
    int cnt = 0;

    std::stack<int> stack;
    stack.push(v);

    while (!stack.empty())
    {
        int v = stack.top();
        stack.pop();
        visited[v] = true;
        cnt += 1;

        for (int nv : adj[v])
        {
            if (!visited[nv])
            {
                stack.push(nv);
            }
        }
    }

    return cnt;
}

int main()
{
    scanf("%d %d", &n, &m);

    for (int i = 0; i < m; i++)
    {
        int a, b;
        scanf("%d %d", &a, &b);
        adj[b].push_back(a);
    }

    int maxCnt = 0;
    for (int i = 1; i <= n; i++)
    {
        ans[i] = dfs(i);
        maxCnt = ans[i] > maxCnt ? ans[i] : maxCnt;
    }

    for (int i = 1; i <= n; i++)
    {
        if (ans[i] == maxCnt)
        {
            printf("%d ", i);
        }
    }
}

File: Baekjoon/1/1325.py
from collections import deque
import sys

n, m = map(int, sys.stdin.readline().split())
adj = [[] for i in range(n+1)]

for i in range(m):
    a, b = map(int, sys.stdin.readline().split())
    adj[b].append(a)


def dfs(v):
    visited = [False for i in range(n+1)]
    stack = deque()
    stack.append(v)

    cnt = 0

    while stack:
        v = stack.pop()
        visited[v] = True
        cnt += 1

        for nv in adj[v]:
            if not visited[nv]:
                stack.append(nv)

    return cnt


ans = [0 for i in range(n+1)]
maxCnt = 0
for i in range(n+1):
    ans[i] = dfs(i)
    maxCnt = max(maxCnt, ans[i])

for i in range(n+1):
    if ans[i] == maxCnt:
        sys.stdout.write(str(i) + " ")

sys.stdout.write('\n')
